genvm/templates/vault/vault.go

```diff
@@ -0,0 +1,72 @@
+package vault
+
+import (
+	"errors"
+	"math/big"
+
+	"github.com/spacemeshos/go-scale"
+
+	"github.com/spacemeshos/go-spacemesh/genvm/core"
+)
+
+var (
+	// ErrNotOwner is raised if Spend is not executed by a principal that matches owner.
+	ErrNotOwner = errors.New("vault: not an owner")
+	// ErrAmountNotAvailable if Spend overlows available amount (see method with the same name).
+	ErrAmountNotAvailable = errors.New("vault: amount not available")
+)
+
+//go:generate scalegen
+
+type Vault struct {
+	Owner               core.Address
+	TotalAmount         uint64
+	InitialUnlockAmount uint64
+	VestingStart        core.LayerID
+	VestingEnd          core.LayerID
+
+	DrainedSoFar uint64
+}
+
+func (v *Vault) isOwner(address core.Address) bool {
+	return v.Owner == address
+}
+
+func (v *Vault) Available(lid core.LayerID) uint64 {
```

Comment:
i know that the research was quite clear about not using layer id to determine the vesting schedule.
to me it's strange to merge this version of vault implementation and change it again later.
but it's fruitless to argue about it now as there is no spec. i guess we'll just merge this and change it later.

Reply:
but are you sure about that? both in Lane's economics implementation and Wojtek's documents layer id is used to determine vesting schedule

it is even discussed in the document that i shared in slack. i know that it is not finalized, but i didn't see even a note that layer id can't be used for vesting anywhere

Reply:
I've never heard any objections for using `LayerID` for the vesting schedule.

The issuance schedule is layer-based. I can't imagine how (or why) to design an issuance schedule that is based on anything else and it doesn't make sense not to tie vesting to the issuance.

@countvonzero do you have any reference to the objection? If it's not written anywhere - who objected? @tal-m?

Reply:
i thought the research want to use TTL to determine whether the vesting account can even start vesting (1 yr),
and not layer ID. or maybe i am confused with not having layer ID available in Verify()? @noamnelke

Reply:
>@countvonzero do you have any reference to the objection? If it's not written anywhere - who objected? @tal-m?

i think it was in one of the video calls where Tal want to use LayerID to determine when the vesting wait time (1 year) has ended and the rest of the research members scoffed at Tal for that, as you seem to just agreed to use TLL and not LayerID for that purpose.